### PR TITLE
GoMod: Fix the algorithm for handling cycles

### LINKED ISSUE #5627: GoMod: Runs out of memory when analyzing the 'ossf/scorecard' project hosted on GitHub
Testing [Scorecard](https://github.com/ossf/scorecard) project ort ran out of heap memory with analyze tool.

command: `./cli/build/install/ort/bin/ort analyze -i ~/scorecard/ -o ortScorecard/`

ort dump:
```
________ _____________________
\_____  \\______   \__    ___/ the OSS Review Toolkit, version 918e64e8eb.
 /   |   \|       _/ |    |
/    |    \    |   \ |    |    Running 'analyze' under Java 17.0.3 on Linux with
\_______  /____|_  / |____|    8 CPUs and a maximum of 5956 MiB of memory.
        \/       \/
Environment variables:
ORT_CONFIG_DIR = /home/alvaro/.ort/config
ORT_DATA_DIR = /home/alvaro/.ort
SHELL = /bin/bash
TERM = xterm-256color

Looking for analyzer-specific configuration in the following files and directories:
	/home/alvaro/.ort/config/curations.yml (does not exist)
	/home/alvaro/.ort/config/curations (does not exist)
	/home/alvaro/scorecard/.ort.yml (does not exist)
	/home/alvaro/.ort/config/resolutions.yml (does not exist)
The following package managers are enabled:
	Bower, Bundler, Cargo, Carthage, CocoaPods, Composer, Conan, DotNet, GoDep, GoMod, Gradle, Maven, NPM, NuGet, PIP, Pipenv, Poetry, Pub, SBT, SpdxDocumentFile, Stack, Unmanaged, Yarn, Yarn2
Analyzing project path:
	/home/alvaro/scorecard
Found 2 GoMod definition file(s) at:
	go.mod
	tools/go.mod
Found 2 definition file(s) from 1 package manager(s) in total.
Aug 01, 2022 4:39:26 PM org.tmatesoft.svn.core.internal.util.DefaultSVNDebugLogger log
SEVERE: WC: Java heap space
java.lang.OutOfMemoryError: Java heap space
	at java.base/java.util.HashMap$KeySet.iterator(HashMap.java:983)
	at org.tmatesoft.svn.core.wc.DefaultSVNRepositoryPool.shutdownInactiveRepositories(DefaultSVNRepositoryPool.java:520)
	at org.tmatesoft.svn.core.wc.DefaultSVNRepositoryPool.access$200(DefaultSVNRepositoryPool.java:69)
	at org.tmatesoft.svn.core.wc.DefaultSVNRepositoryPool$TimeoutTask.run(DefaultSVNRepositoryPool.java:534)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

16:39:26.629 [DefaultDispatcher-worker-1] ERROR org.ossreviewtoolkit.analyzer.managers.GoMod - Resolving GoMod dependencies for path 'go.mod' failed with: OutOfMemoryError: Java heap space

```

### PR DESCRIPTION
**Pull request by fviernau**
The previous algorithm exhibits the following issues when run on graphs
with cycles, in particular when the amount of edges is large:

1. Cycles of length 1 lead to infinite recursion.
2. The algorithm is inefficient in terms of execution time and memory
   allocation, as it creates a copy of the
   `predecessorNodes` set per recursion.
3. When run against [1] the execution of `toPackageReferenceForest()`
   didn't finish within 15 minutes, causing high CPU load and memory
   consumption.

If GoMod used the dependency graph format already, the solution would be
as simple as calling `DependencyGraphBuilder.breakCycles()`. Fix the
problem by copying the code of `DependencyGraphBuilder.breakCycles()` as
a temporary quick fix. This saves effort, because refactoring GoMod to
use the dependency graph is planned anyway [2], which will allow for
removing that copied code again.

[1] https://github.com/ossf/scorecard
[2] https://github.com/oss-review-toolkit/ort/issues/4249

Fixes https://github.com/oss-review-toolkit/ort/issues/5627.

### CODE REVIEW

**Comment by dgutson** [comment]
Thanks @fviernau!, is there an issue already created to "refactor GoMod to
use the dependency graph"?

**Comment by qequ** [comment]
Tested the analyze tool with the [Scorecard repo](https://github.com/ossf/scorecard) and works fine without the need of adding extra memory 👌

File: analyzer/src/main/kotlin/managers/GoMod.kt [hunk 1]
```diff
@@ -373,14 +374,60 @@ private class Graph(private val nodeMap: MutableMap<Identifier, Set<Identifier>>
...
+            visit(v)
+        }
+
+        return Graph(outgoingEdgesForNodes.mapValuesTo(mutableMapOf()) { it.value.toSet() })
+    }
+
     /**
      * Convert this [Graph] to a set of [PackageReference]s that spawn the dependency trees of the direct dependencies
-     * of the given [root] package.
+     * of the given [root] package. The graph must not contain any circles, so [breakCycles] should be called before.
```
Comment (oheger-bosch) [inline analyzer/src/main/kotlin/managers/GoMod.kt:425]: Should this be "...must not contain any cycles..." instead of "circles"?

**Review by oheger-bosch — APPROVED** [review]

**Comment by tsteenbe** [comment]
> Thanks @fviernau!, is there an issue already created to "refactor GoMod to use the dependency graph"?
@dgutson  Yes, there is see https://github.com/oss-review-toolkit/ort/issues/4249

**Review by oheger-bosch — DISMISSED** [review]

**Review by oheger-bosch — APPROVED** [review]

**Pull request merged**